docs/source/reference/ext.rst

```diff
@@ -19,13 +19,15 @@ CuPy-specific functions are placed under ``cupyx`` namespace.
 DLPack utilities
 ----------------
 
-This is a helper function for creating a :class:`cupy.ndarray` from a DLPack tensor.
+Below are helper functions for creating a :class:`cupy.ndarray` from either a DLPack tensor
+or any object supporting the DLPack data exchange protocol.
 For further detail see :ref:`dlpack`.
 
 .. autosummary::
    :toctree: generated/
 
    cupy.fromDlpack
+   cupy.from_dlpack
```

Comment:
Honestly speaking I don't want to document this API, as it's likely to be removed in the future & having two similar API is confusing to users. How about changing `cupy.fromDlpack` to support both raw pycapsule and `__dlpack__` protocol?
@emcastillo is working in this approach in PyTorch: https://github.com/pytorch/pytorch/pull/57110

Reply:
This API is [documented](https://data-apis.org/array-api/latest/API_specification/creation_functions.html#from-dlpack-x) as part of the Array API standard. People have invested great effort into it, so I don't think it'd go away 🙂 

I would actually suggest to deprecate `cupy.fromDlpack` as it is known unsafe w.r.t. non-default streams
https://github.com/cupy/cupy/blob/2f8552bade9f5994a9417922da272778dcf0c850/docs/source/user_guide/interoperability.rst#L296
It is functionally equivalent to `arr.__dlpack__(None)` so users do not lose anything. With more number of libraries supporting the new DLPack protocol (`__dlpack_device__` + `__dlpack__` + `from_dlpack()`) such as us, NumPy, PyTorch, RAPIDS, etc, I think it is better to encourage users to switch to the new protocol and not manage the pycapsule object by themselves.

As for making `from_dlpack()` support both pycapsule and compliant arrays, I feel it's best that we discuss it in the PyTorch PR (or the array-api repo) first, and then implement the decision. I personally think this is confusing and against the safe design of the protocol...

Reply:
First of all, I really thank many people, including you, for the efforts on standardization! 😄 
My understanding is that we will introduce `cupy.array_api.from_dlpack()` at some point and then remove `cupy.from_dlpack()` and `cupy.fromDlpack()`. Is this correct?

> I would actually suggest to deprecate cupy.fromDlpack as it is known unsafe w.r.t. non-default streams

I agree with deprecating `fromDlpack()`, however as most existing producers do not support `__dlpack__` yet, we still have to provide a way to load raw PyCapsule objects. So I think supporting both raw PyCausule and `__dlpack__` in `from_dlpack` is a good tentative solution for now.

Reply:
> First of all, I really thank many people, including you, for the efforts on standardization! 😄

My pleasure! I was just trying to stress that the hope is to make the DLPack protocol sustain. We can definitely mark it experimental at least in CuPy v10, given that there are still ongoing discussions (ex: https://github.com/dmlc/dlpack/issues/74), but the goal is to improve it, not to replace it by something else.

> My understanding is that we will introduce `cupy.array_api.from_dlpack()` at some point and then remove `cupy.from_dlpack()` and `cupy.fromDlpack()`. Is this correct?

It depends. There is a [discussion](https://github.com/cupy/cupy/issues/4789#issuecomment-864725951) I'd like to kick off with the team. If we follow NumPy and have two separate namespaces, `cupy.from_dlpack()` can stay (and return a `cupy.ndarray`) and a new `cupy._array_api.from_dlpack()`will be added (and return a `cupy._array_api.ndarray`). 

In any case I hope to deprecate and remove `cupy.fromDlpack()`.

> So I think supporting both raw PyCausule and `__dlpack__` in `from_dlpack` is a good tentative solution for now.

Sounds like this will be [an acceptable workaround even in PyTorch](https://github.com/pytorch/pytorch/pull/57110#issuecomment-864599710), so I will follow!

Reply:
> If we follow NumPy and have two separate namespaces, `cupy.from_dlpack()` can stay (and return a `cupy.ndarray`) and a new `cupy._array_api.from_dlpack()`will be added (and return a `cupy._array_api.ndarray`). 

I see, sounds nice.

> In any case I hope to deprecate and remove `cupy.fromDlpack()`.

Agree.

> Sounds like this will be [an acceptable workaround even in PyTorch](https://github.com/pytorch/pytorch/pull/57110#issuecomment-864599710), so I will follow!

:+1:

Reply:
@kmaehashi I am wondering if we should deprecate `toDlpack` as well...

Reply:
I think `toDlpack` deprecation should happen after most consumers support new `from_dlpack` protocol.